File: drills/ts/solutions/nc71_coin_change.ts
export function coinChange(coins: number[], amount: number): number {
  // Build up from zero: the cheapest way to make a target is one coin more than
  // the cheapest way to make what is left after removing some coin. An amount
  // with no entry is simply unreachable, which saves inventing a sentinel for
  // infinity.
  const table = new Map<number, number>([[0, 0]]);

  for (let target = 1; target <= amount; target++) {
    const options = coins
      .filter((coin) => coin <= target && table.has(target - coin))
      .map((coin) => table.get(target - coin)!);
    if (options.length) table.set(target, Math.min(...options) + 1);
  }

  return table.get(amount) ?? -1;
}

File: drills/go/solutions/nc71_coin_change.go
package main

func coinChange(coins []int, amount int) int {
	// fewest[a] is the fewest coins making a; a sentinel above any real
	// answer stands for "not yet reachable".
	unreachable := amount + 1
	fewest := make([]int, amount+1)
	for a := 1; a <= amount; a++ {
		fewest[a] = unreachable
		for _, coin := range coins {
			if coin <= a {
				fewest[a] = min(fewest[a], fewest[a-coin]+1)
			}
		}
	}
	if fewest[amount] == unreachable {
		return -1
	}
	return fewest[amount]
}
